### PR TITLE
feat(osv) : add osv support

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
Support OSV format
https://ossf.github.io/osv-schema/#python-vulnerability

## Supported languages
- Go
- Python
- Rust

### CODE REVIEW

File: osv/types.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+// type OsvDatabase struct {
+// }
+
+type OsvReference struct {
+	Type string `json:"type,omitempty"`
+	Url  string `json:"url,omitempty"`
+}
+
+//collection based on https://ossf.github.io/osv-schema/
+type OsvJson struct {
```
Comment (AndreyLevchenko) [inline osv/types.go:37]: `Json` postfix isn't very good choice as structure can be used some way which is not related to json serialization

File: osv/osv.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	url           string
+	dir           string
+	ecosystemDirs map[string]string
+}
+
+type option func(*options)
+
+type Osv struct {
+	opts  *options
+	AppFs afero.Fs
```
Comment (AndreyLevchenko) [inline osv/osv.go:38]: `AppFs` is not required here

File: osv/osv.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+func WithEcosystem(ecosystemDir map[string]string) option {
+	return func(opts *options) {
+		opts.ecosystemDirs = ecosystemDir
+	}
+}
+
+func NewOsv(opts ...option) Osv {
+	ecosystemDirs := make(map[string]string)
```
Comment (AndreyLevchenko) [inline osv/osv.go:60]: this should be used only if `WithEcosystem()` hasn't been called.

File: osv/osv_test.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+			for _, ecosystem := range tt.ecosystem {
+				testEcosystemDir[ecosystem.name] = ecosystem.dir
+			}
+
+			c := osv.NewOsv(osv.WithURL(testUrl), osv.WithDir(testDir), osv.WithEcosystem(testEcosystemDir))
+			err := c.Update()
+			if tt.wantErr != "" {
+				require.Error(t, err)
+				assert.Contains(t, err.Error(), tt.wantErr)
+				return
```
Comment (AndreyLevchenko) [inline osv/osv_test.go:92]: I would better use `else`

**Review by AndreyLevchenko — CHANGES_REQUESTED** [review]

File: osv/osv_test.go (hunk 4 above)
Comment (DmitriyLewen) [reply osv/osv_test.go]: done

File: osv/osv.go (hunk 3 above)
Comment (DmitriyLewen) [reply osv/osv.go:68]: added check

File: osv/osv.go (hunk 2 above)
Comment (DmitriyLewen) [reply osv/osv.go]: removed

File: osv/types.go (hunk 1 above)
Comment (DmitriyLewen) [reply osv/types.go:39]: changed

File: osv/osv.go (hunk 3 above)
Comment (AndreyLevchenko) [reply osv/osv.go:68]: Just call all `options` and then check if `ecosystemDirs` is still empty. If it is then assign defaults

Comment (DmitriyLewen) [reply osv/osv.go]: changed

File: osv/osv.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+		o.dir = filepath.Join(utils.VulnListDir(), osvDir)
+	}
+
+	return Osv{
+		opts: o,
+	}
+}
+
+func (osv *Osv) Update() error {
+	allfiles, err := osv.getAllFiles()
```
Comment (knqyf263) [inline osv/osv.go:87]: I prefer iterating the ecosystem here and downloading files inside each iteration.

```
for ecoSystem, ecoSystemDir := range osv.opts.ecosystemDirs {
	log.Printf("Updating OSV %s advisories", ecoSystem)
	downloadFiles(ecoSystem, ecoSystemDir)

	err := filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
		f, err := os.Open(path)
		json.NewDecoder(f).Decode()...
	}
}
```

File: osv/osv.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+	if err != nil {
+		return xerrors.Errorf("failed to get files: %w", err)
+	}
+
+	bar := pb.StartNew(osv.getAmountFiles(allfiles))
```
Comment (knqyf263) [inline osv/osv.go:92]: I think it is fast enough. No need to show a progress bar.

File: osv/types.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
+package osv
+
+type OsvAffected struct {
+	Package  *OsvPackage `json:"package,omitempty"`
+	Ranges   []OsvRange  `json:"ranges,omitempty"`
+	Versions []string    `json:"versions,omitempty"`
+
+	// Ecosystem *OsvEcosystem `json:"ecosystem_specific,omitempty"` //The meaning of the values within the object is entirely defined by the ecosystem
```
Comment (knqyf263) [inline osv/types.go:8]: You can use `interface{}`

```suggestion
	EcosystemSpecific interface{} `json:"ecosystem_specific,omitempty"`
```

File: osv/types.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+	// Database  *OsvDatabase  `json:"database_specific,omitempty"` //The meaning of the values within the object is entirely defined by the database
```
Comment (knqyf263) [inline osv/types.go:10]: ditto

File: osv/types.go (hunk 7 above)
Comment (DmitriyLewen) [reply osv/types.go]: Done

File: osv/types.go (hunk 8 above)
Comment (DmitriyLewen) [reply osv/types.go]: done

File: osv/osv.go (hunk 5 above)
Comment (DmitriyLewen) [reply osv/osv.go]: changed

File: main.go [hunk 9]
```diff
@@ -193,6 +194,12 @@ func run() error {
 			return xerrors.Errorf("AlmaLinux update error: %w", err)
 		}
 		commitMsg = "AlmaLinux Security Advisory"
+	case "osv":
+		p := osv.NewOsv()
+		if err := p.Update(); err != nil {
+			return xerrors.Errorf("error in Osv update: %w", err)
+		}
+		commitMsg = "Osv Security Advisories"
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply main.go]: Done

File: osv/osv.go (hunk 6 above)
Comment (DmitriyLewen) [reply osv/osv.go]: Removed

**Review by knqyf263 — APPROVED** [review]

**Comment by knqyf263** [comment]
Thanks

**Pull request merged**